### PR TITLE
Merge advanced analytics tests to gpdb5 master pipeline

### PR DESCRIPTION
**Pull request by jingyimei**
Both madlib and postgis integration tests are added to gpdb5 master pipeline.

### CODE REVIEW

**Comment by Eulerizeit** [comment]
Does it make sense to add tests for code outside the repo to a repo? I
guess what I'm asking is, maybe it is time to move concourse to its own
repo so we don't have confusion.

Rob

**Comment by danielgustafsson** [comment]
I think we need to be careful to add dependencies on external projects outside our control without first defining the “social contract” for these. If either of these repos have a bug making our pipeline red, we are blocked from going forward without a clear definition on who is responsible for fixing it.

To me it would make more sense to have a completely separate integration pipeline which tests GPDB with various external projects, while keeping the GPDB pipeline focused on GPDB (where we have a clear definition of what Read means etc).

File: concourse/pipelines/pipeline.yml [hunk 1]
```diff
@@ -274,21 +302,6 @@ resources:
       ###########################################################
       bucket_path: {{tf-bucket-path}}
 
-- name: terraform_for_dpm
```
Comment (chrishajas) [inline concourse/pipelines/pipeline.yml]: This is used in the DPM tests I believe.

**Comment by chrishajas** [comment]
It feels a bit weird having this in the main GPDB pipeline since it isn't present in the gpdb repo. If madlib is simply a consumer of greenplum (just relies on the binary), it might make more sense to have a separate integration pipeline. That way, if a madlib test goes red, it doesn't stop the main GPDB pipeline. 

We do a similar thing in gpbackup. We have a separate pipeline that consumes the gpdb binaries that are generated in the main pipeline. We then have separate tests that test gpbackup on greenplum. If there is any madlib test that would break as a result of a greenplum change (as opposed to a madlib change), that specific test should be added to ICG as it is showing a gap in ICG coverage.

File: concourse/pipelines/pipeline.yml (hunk 1 above)
Comment (jingyimei) [reply concourse/pipelines/pipeline.yml]: Thanks for the reminder, it may caused by wrong rebase. Updated.

**Comment by jingyimei** [comment]
In those jobs we use last release version of MADlib and postgis gppkg to test against latest gpdb binary, and the main idea here is to detect new GPDB changes that break MADlib and postgis. When somebody make new commit to gpdb, it is easier for them to notice the new commit breaks something on madlib/postgis from 5X_STABLE and 5_master pipeline instead of our separate integration pipeline. It is also more convenient for team collaboration to fix the bug. @pivotal-mike @edespino  Any thoughts about this?

**Comment by njayaram2** [comment]
To add to @jingyimei's comment, since the latest commits to GPDB (stable, master and 4.3X) are tested against a non-changing MADlib/PostGIS artifact (latest stable releases of the same), the most likely reasons for these jobs going red is some change in the GPDB code base. The plan is to apply the guidelines we have for pushing code when gpdb pipelines are red for these jobs too. The MADlib team will work with relevant teams to reproduce the issue.
Of course, there could be some obscure unknown MADlib issue which might cause the jobs to go red when some GPDB code changes, but that is expected to be less likely. If that happens, we will surely work with the relevant teams to resolve it.

**Comment by ashwinstar** [comment]
On Mon, Oct 16, 2017 at 9:34 AM, Nandish Jayaram <notifications@github.com>
wrote:

> To add to @jingyimei <https://github.com/jingyimei>'s comment, since the
> latest commits to GPDB (stable, master and 4.3X) are tested against a
> non-changing MADlib/PostGIS artifact (latest stable releases of the same),
> the most likely reasons for these jobs going red is some change in the GPDB
> code base. The plan is to apply the guidelines we have for pushing code
> when gpdb pipelines are red for these jobs too. The MADlib team will work
> with relevant teams to reproduce the issue.
> Of course, there could be some obscure unknown MADlib issue which might
> cause the jobs to go red when some GPDB code changes, but that is expected
> to be less likely. If that happens, we will surely work with the relevant
> teams to resolve it.
>
There are various other partners and tools which need to work with GPDB, we
won't be brining in each of those inside GPDB pipeline to verify. This
completely reverses the model. To re-iterate if there are any failures
detected by users of GPDB which are not caught by current tests we need to
enhance coverage of GODB from that perspective when found, but do not agree
to have not in repo tests added to main pipeline.

**Comment by fmcquillan99** [comment]
Postgis IC & madlib IC & madlib Tinc are all release blockers for GPDB, so we should ensure that we have proper visibility of the state of those tests across all teams, and not just the advanced analytics team.  With separate pipelines for postgis and madlib, will we achieve that desired visibility? And will we respond in the right timeframe to breaking tests?

**Comment by pivotal-mike** [comment]
It should be noted that there are a number of use cases/code paths that both MADlib and postGIS exercise that are not being exercised in our normal tests.  While the ideal situation would be to have these use cases captured in ICW or isolation2, both MADlib and postGIS are helping to extend the code coverage for server at this time.  For that, if for no other reason I'm going to give this a +1

**Comment by danielgustafsson** [comment]
> On 16 Oct 2017, at 21:19, Michael Roth <notifications@github.com> wrote:
> 
> It should be noted that there are a number of use cases/code paths that both MADlib and postGIS exercise that are not being exercised in our normal tests. While the ideal situation would be to have these use cases captured in ICW or isolation2, both MADlib and postGIS are helping to extend the code coverage for server at this time. For that, if for no other reason I'm going to give this a +1

Having done more thinking after my initial comment here, the way I see it is that we are adding test coverage to address holes in the coverage of GPDB. This happens to be a real testsuite from Madlib, but it could just as well have been renamed gobbledygook since we are testing GPDB and not Madlib.

That being said, I wholeheartedly agree with previous comments that we need to increase coverage of GPDB so that we can avoid the need for this. What we need to do is for every bug found with this, extend coverage with the aim to be able to remove this again.

**Comment by Eulerizeit** [comment]
On Mon, Oct 16, 2017 at 8:00 PM, Daniel Gustafsson <notifications@github.com
> wrote:
>
>
> That being said, I wholeheartedly agree with previous comments that we
> need to increase coverage of GPDB so that we can avoid the need for this.
> What we need to do is for every bug found with this, extend coverage with
> the aim to be able to remove this again.
>

+1 Can we make that an explicit goal?

**Comment by cjcjameson** [comment]
My view would be that it's great to run Madlib, postgis, or other external collaborators within our pipeline, and OK to have our master pipeline break or go red, on the following conditions:

1. We explicitly pin to a GA release version of the external collaborator. We should hard code that in the pipeline, rather than having it be a value that is interpolated. Bumping the version against which we test is a simple, but important, PR to make each time.
2. The tests and test infrastructure/runner code for the external collaborator is versioned with the source code of the external collaborator.
3. The external collaborator's codebase + tests + test infrastructure are open source
4. Using this, we specifically use the same pin or an equivalent pin-reference to refer to a tagged commit in their repo. We pull in their testing code. We may provide some task.ymls perhaps in the GPDB codebase to ingest their testing, if they do not use concourse, but we could also ask that they codify how to run their tests in the Concourse task.yml format as well.

The essence of my argument is that these are collaborators which might strive to give us regular feedback about where we need more test coverage, but in reality, it's difficult. And bringing the testing closer together will help with faster feedback.

The GPORCA project / codebase is a very special case, but adheres to the above pattern for the most part.

If there were closed-source collaborators, then it would be entirely on a closed source contributor to write contract-type tests if we wanted to assert things about them.

Is part of the hesitation to expand the pipeline based on infrastructure instability or general flakiness -- fear that if it gets too big, it's too likely that we'll be slowed down by sporadic failures?

**Comment by Eulerizeit** [comment]
On Tue, Oct 17, 2017 at 1:46 AM, C.J. Jameson <notifications@github.com>
wrote:

> <SNIP>
>
> Is part of the hesitation to expand the pipeline based on infrastructure
> instability or general flakiness -- fear that if it gets too big, it's too
> likely that we'll be slowed down by sporadic failures?
>

For me the hesitation is that commits should stop when the CI goes red. The
more tests (especially on a notoriously shaky CI) the higher the likelihood
of a red CI.

Fundamentally we should be testing stuff that happens inside the DB, MADlib
and PostGIS both fit this actually.

Ideally we would have a way to test the APIs and unit test the code for the
UDFs prior to pushing the code into the main repo so that we could feel
secure that the issue was a Database problem or a subtile breaking of the
API. This would mean fully testing the code outside of the master pipeline
before committing it and then only testing Green builds on green builds. We
ought to be doing this with everything not inside the DB BTW.

-- Rob

**Comment by chrishajas** [comment]
If this is enhancing code coverage of GPDB, then I agree that it should go in this suite. In that case, I feel that the madlib/postgis binaries we test against should be static and require a commit to the pipeline file in this repo rather than the secrets file to update the binary versions.

**Comment by jingyimei** [comment]
@chrishajas That is a great point, will update the PR to use a static gppkg name for both MADlib and PostGIS.

**Comment by pivotal-mike** [comment]
@jingyimei I think we've kicked the tires enough on this and the tests are live on the pipeline.  Can we close/merge this please?

**Comment by jingyimei** [comment]
@pivotal-mike Those tests are broken by several bugs and QP have multiple stories ongoing to fix them. After discussion with QP, we decided to wait til the end of next week to merge them. Besides, this pipeline is blocked by gpdb version bumping to 6.0 and we have a story to generate new madlib gppkg that supports the new version.This story should also be done by next week.

**Comment by kaknikhil** [comment]
@pivotal-mike we will not be able to merge this pull request by the end of this week. One of the failures is because of a bug in madlib and postgis and we may not be able to get to it by this week. https://github.com/greenplum-db/gpdb/pull/3621

**Comment by ashwinstar** [comment]
@jingyimei  Is this PR going any further or should be closed ?

**Comment by ashwinstar** [comment]
@jingyimei @kaknikhil is this moving further or should be closed ?

**Comment by kaknikhil** [comment]
@ashwinstar Thanks for the reminder,we will close this pull request for now.  
MADlib and postgis were broken because of some bugs in gpdb master, some of which have been fixed but not all. We will need to test MADlib/postgis with gpdb master again to get a list of all the failures and then create issues/tracker stories. We will reopen the pull request once all these issues have been fixed.

**Pull request closed without merging**